Feature/custom image block

AshTNA opened the pull request:
Hi @gtvj @jacobtoppm

Would you be able to look at this PR? It fixes #175 

It includes the changes to implement a custom image block in Wagtail which ensures accessibility is reflected on the page.

I have replaced the two `ImageChooserBlock`s on the Insights page instead of adding alt_text fields to this block, to prevent code duplication - this will require editors to reintroduce the images to the LIVE site but it will ensure the new ImageBlock is being used everywhere it is needed. **EDIT: We now know that this can be achieved with a data migration.** 

One thing I couldn't figure out was how to make the `hero_image` use the new `ImageBlock`, as it is implemented via a Foreign Key and seems to be using a `ImageChooserPanel` (see `etna/heroes/models.py`). Therefore I've removed the alt text from `templates/includes/hero-img.html` completely but @gtvj let me know if you think this needs to use the `ImageBlock`. **EDIT: Jacob helped to fix this!**

The editorial has been approved by MDV. 

To test the ImageBlock, create an Insights Page locally, and then include a `Featured Record` or `Promoted Item` on the page.

Thanks :+1:

![image](https://user-images.githubusercontent.com/8880610/138439365-6b377d55-d96a-481a-97e2-116ea16daf66.png)

gtvj commented:
Hi @AshTNA - there are a few conflicts here and I'm just reflecting on the issue we had with CI this morning. Could you rebase onto `main`, fix any conflicts and check that CI is happy?

AshTNA commented on etna/heroes/forms.py:
```diff
@@ -0,0 +1,36 @@
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.fields['hero_image_decorative'].label = format_html("%s <p style='font-size: 11px;'>%s</p>" % (
```
Thanks @jacobtoppm - @gtvj do you think we should incorporate this now, or add it as a task for later so that we have a process for adding TNA specific styles to the dashboard?

gtvj replied:
Hi @AshTNA - I think it would be best to do it now. Looks quite straightforward (on the face of it).

AshTNA replied:
Thanks @gtvj - I've done this, I had to create a separate SASS file otherwise all of our toolkit styles would ruin the dashboard CSS. I've added this new file to the readme, and the deployment config file.

AshTNA commented:
Thanks @gtvj - I've rebased now :+1:

@jacobtoppm I've started to getting some linting errors on this PR - do all of these need to be resolved for the CI to complete? Just curious as some of the linting issues it is complaining about seem to exist on the `main` branch too.

There is also a failing test at `etna/records/tests/test_blocks.py` on the `test_add_featured_record` test. I'm guessing this is due to the new `teaser_image` block being a part of that page and having validation requirements? Thanks

AshTNA commented:
@jacobtoppm I've fixed all the linting issues apart from the "line too long" ones - would you know how we could resolve those? Thanks :+1:

jacobtoppm commented:
@AshTNA There are a few ways you can do multiline strings in python. If you want your line breaks when to write it to be converted to newlines in the string, you can enclose your string in triple quotes. If not, use parentheses to divide a string across lines as so (note the lack of commas):

```
('my'
'long'
'string')
```

or similar

jacobtoppm commented:
@AshTNA Also, to answer your earlier question, yes, all linting errors need to be resolved for CI to complete. I would suggest making sure all CI passes before merging, as otherwise several people can end up needing to make fixes in their own branches, leading to confusion.

The test is interesting, but your suggestion sounds like a likely cause - consider sticking a debug statement in (`breakpoint()` in the test code) and running the tests locally `docker-compose exec web bash` then `poetry run ./manage.py test`, then using the debugger to step through to find out what's going on and investigate any variables of interest, what the response looks like for instance, and whether the page was actually saved to the db. Let me know if you have any trouble and want to talk it though!

jacobtoppm commented:
@AshTNA Definitely, it's something I raised as a future option when we initially added CI. Black is a very nice autoformatter. The one thing I would say is that if you start using black, you should probably:
- make sure everyone is using it
- add it as an additional CI check
As it is very opinionated as a formatter (much more so than flake8, our CI check), and otherwise, when someone who is using it runs it, it will reformat a lot of code, resulting in periodic big reformatting commits.

If everyone's okay with adding it to the workflow though, it could be a very nice addition for speeding up formatting.

AshTNA commented:
Thanks @jacobtoppm - I'm learning a lot here! I'm getting different test errors locally to the ones GitHub is throwing, getting a bit confused so  I may need some help fixing the tests. No worries if you are too busy at the moment.

AshTNA commented:
This is two of the three test errors:

- `File "/app/etna/records/tests/test_blocks.py", line 179, in test_view_edit_page_with_featured_record`
- `File "/app/etna/records/tests/test_blocks.py", line 359, in test_view_edit_page_with_featured_record`

The error is:

```python
    self.assertEqual(len(responses.calls), 1)
AssertionError: 2 != 1
````

It looks like `responses.calls` has two responses in it that are the same:

```python
(Pdb) responses.calls[0].request.url
'https://kong.test/data/fetch?iaid=C123456&from=0&pretty=false&expand=false'
(Pdb) responses.calls[1].request.url
'https://kong.test/data/fetch?iaid=C123456&from=0&pretty=false&expand=false'
``` 

 Is this an issue with the test itself, or is it something I've introduced in this PR?

jacobtoppm commented:
@AshTNA So far, the issue I am seeing is with the `assertRedirects` - no other. Looking at it in the debugger, it seems that it's to do with the `FeaturedRecordBlock`. Because you have added `ImageBlock` as a sub-block in the StructBlock, one automatically gets created for every featured record. Because it is invalid to have an `ImageBlock` with no `image`, the subblock, `ImageBlock`, then fails its own validation.

So basically, the teaser image is currently not actually optional. (`required=False` doesn't mean you can add the sub-block or not, it always gets added, it's about whether an empty value is acceptable)

Options:
- Use `required` on `ImageBlock` to determine whether `image` is required or not. If there is no image, hide the block on the FE
- Wrap `ImageBlock` within a `StreamBlock` or similar inside `FeaturedRecordBlock` to allow optionally creating a sub-block or not.

AshTNA commented:
Thanks @jacobtoppm - I tried to set the image in `ImageBlock` to `required=False`, however I was still getting issues. 

@jacobtoppm @gtvj  I think there may be a bigger issue with the repo as I'm getting this issue on the UI on both `main`  and on this branch. 

This is what I see on  `main`. I get the same on this branch and none of my custom validation messages show up.

![image](https://user-images.githubusercontent.com/8880610/142197153-604008b1-37de-4b72-af3b-03924ae4a99d.png)

![image](https://user-images.githubusercontent.com/8880610/142197227-573b6873-bac9-43d5-ab10-fa7c46785aa6.png)

jacobtoppm commented:
Hi @AshTNA - this is likely due to Mo's Wagtail update PR being merged in the meantime, which changed how the Wagtail admin renders streamfield widgets. Have you tried:
- running `poetry install` within the container/rebuilding the web container
- rebuilding the FE
- clearing your browser cache?

AshTNA commented:
Thanks @jacobtoppm - This was on a fresh instance of Wagtail, as I deleted my docker instance and created a fresh one when I switched to `main` - I've double checked the dashboard in Incognito mode in case it was the cache but the issue still persists. Thanks :+1:

jacobtoppm commented:
Hi @AshTNA  - I'm not sure what issue you're running into with the choosers, but the fontawesome breaking is likely due to the change to svg for streamfield icons. There should be a dropin replacement package we can install: https://pypi.org/project/wagtail-font-awesome-svg/

AshTNA commented:
Thanks @jacobtoppm - removing fontawesome from the installed apps removed the JS error.

@gtvj who should be responsible for reinstalling fontawesome? I can try but don't have much experience working with poetry etc.

@jacobtoppm I will have to investigate some more as my ImageBlock seems to have lost its validation errors. The page still says there are validation errors, but they don't appear next to the fields with their relevant error message. Also, `format_html` has stopped formatting the HTML so I'm not sure what's going on with the block!

jacobtoppm commented:
Hi @AshTNA, @gtvj - depending who you want to do this, I would be happy to do it myself or support Ash with trying out poetry, as this should be (fingers crossed) a fairly straightforward replacement.

Thanks for flagging, Ash! Not sure what's going on with `format_html` yet, but for the validation issues, check out this issue: https://github.com/wagtail/wagtail/issues/7202 - see if the same fix works for you?

AshTNA commented:
Thanks @jacobtoppm - changing it to a `StructBlockValidationError` fixed the issue. I'll try and investigate the tests and `format_html`.

AshTNA commented:
Thanks @jacobtoppm for your help, all things seem to be working now!

@gtvj - we will need to look into creating the data migration for this at some point.

jacobtoppm commented:
No problem!

@gtvj @AshTNA We did discuss that you might not want to do a data migration, but just guide editors into filling in the replacement images + alt text as they would need to add alt text anyway when this is deployed. I'm not sure where you stand on this now?

AshTNA commented:
I believe the idea was that we should go with a data migration, so that we have experience of completing one for future reference. It's likely in the future we'd need to conduct a data migration!

ababic commented on etna/media/blocks.py:
```diff
@@ -1,9 +1,21 @@
 from wagtailmedia.blocks import AbstractMediaChooserBlock
 
 
+@hooks.register("insert_editor_css")
+def editor_css():
+    return format_html(
+        '<link rel="stylesheet" href="{}">', static("/css/dist/etna-wagtail-editor.css")
+    )
+
+
```
Wagtail hooks should always be registered in a `wagtail_hooks.py` file in an app where possible, as Wagtail always looks for them in that location. I would suggest moving this code there to make things easier to find for developers coming from other Wagtail projects.

AshTNA replied:
Thanks @ababic - that makes sense!

ababic commented:
@AshTNA For the sake of overriding a single style for the Wagtail admin, I wonder whether it's worth having that as a `sass` file that needs to be compiled. If it were just a `.css` file, you wouldn't need that step, and wouldn't need to update the README/docs either.

AshTNA commented:
> @AshTNA For the sake of overriding a single style for the Wagtail admin, I wonder whether it's worth having that as a `sass` file that needs to be compiled. If it were just a `.css` file, you wouldn't need that step, and wouldn't need to update the README/docs either.

Thanks @ababic - I think it's best to keep it in SASS for future proofing as I suspect it could be possible that we implement some BEM CSS for the admin pages, and that's a lot easier to achieve with SASS!

AshTNA commented:
Hi @ababic - thanks for doing the data migration. As we know the teaser images were disappearing locally which we thought might be because of platform.sh. However, the images that are missing after the data migration seem to exist on my local version of the `main` branch, so I do believe they are being pulled down by platform.sh and added to the media database. Therefore the data migration may be the issue. If possible would you be able to investigate this?

Thanks :+1:

ababic commented:
@AshTNA The data migration has now been tweaked to account for the changes in block names, e.g. `teaser_image` -> `image.image` and `teaser_alt_text` -> `image.alt_text`

AshTNA commented:
@ababic It should be OK to merge, if you're happy with the code :+1:
